### PR TITLE
Rewrite job runner in Rust

### PR DESCRIPTION
**Pull request by NathanFlurry**
## Summary

Replaces the old run.sh with a new Rust-based binary.

## Motivation

The old run.sh script had the following problems:

- Did not handle SIGTERM and pass it to the runc container correctly
- Piping logs to Vector was incredibly hacky
- Log shipping was flakey
- Spawning another instance of Vector for log shipping required extra resources

After a few attempts to fix these issues, it became increasingly difficult to do this reliably in Bash and the magical-random-symbol-toxic-spew became worse and worse.

## Other benefits

This architecutre has other added benefits:

- Very easy to test thoroughly
- Rate limiting is done as soona s possible
- We now log a custom message when lines are rate limited or truncated (which is not possible with Vector)
- The setup_* scripts can be converted to Rust eventually for better testing and reliability

### CODE REVIEW

**Comment by NathanFlurry** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/332?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #324** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/324?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #314** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/314?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #310** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/310?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #311** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/311?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #312** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/312?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #313** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/313?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #315** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/315?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #316** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/316?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #326** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/326?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #325** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/325?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                    * **PR #329** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/329?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                      * **PR #330** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/330?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                        * **PR #331** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/331?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                          * **PR #332** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/332?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
                            * **PR #333** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/333?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                              * **PR #334** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/334?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/rivet-gg/rivet/332?utm_source=stack-comment).

**Comment by NathanFlurry** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/332?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #324** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/324?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #314** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/314?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #310** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/310?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #311** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/311?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #312** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/312?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #313** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/313?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #315** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/315?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #316** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/316?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #326** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/326?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #325** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/325?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                    * **PR #329** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/329?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                      * **PR #330** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/330?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                        * **PR #331** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/331?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                          * **PR #332** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/332?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
                            * **PR #333** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/333?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                              * **PR #334** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/334?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/rivet-gg/rivet/332?utm_source=stack-comment).

**Comment by NathanFlurry** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/332?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #324** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/324?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #314** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/314?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #310** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/310?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #311** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/311?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #312** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/312?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #313** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/313?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #315** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/315?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #316** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/316?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #326** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/326?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #325** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/325?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                    * **PR #329** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/329?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                      * **PR #330** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/330?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                        * **PR #331** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/331?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                          * **PR #332** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/332?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
                            * **PR #333** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/333?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                              * **PR #334** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/334?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/rivet-gg/rivet/332?utm_source=stack-comment).

File: lib/job-runner/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+		// much can be logged per minute. This is here to prevent games
+		// that log as fast as possible (i.e. positions of objects every
+		// tick) from exhausting the system while still allowing sane
+		// amounts of logging. This happens very frequently.
+		//
+		// 4 logs/s * 1024 bytes/log = 4096 bytes/lobby/s = 14.7 MB/lobby/hr = 353.8 MB/lobby/day  = 10.6 GB/lobby/month
+		let mut throttle_long = throttle::Throttle::new(1200, Duration::from_secs(300));
+
+		// Throttles error logs
+		let mut throttle_error = throttle::Throttle::new(1, Duration::from_secs(60));
```
Comment (ABCxFF) [inline lib/job-runner/src/main.rs:151]: I feel like this throttle might too much; What if for some reason its an error with a stack trace gets printed to stderr, would be nice to get the full stack trace.

Maybe something like 50 every 10 minutes?

Comment (NathanFlurry) [reply lib/job-runner/src/main.rs:151]: `throttle_error` is different than stderr, this is only for limiting how frequently we log the rate limit error internal. Both stdout and stderr have the same throttle.

Comment (ABCxFF) [reply lib/job-runner/src/main.rs:151]: Ah okay, I see now, makes sense

**Review by MasterPtato — APPROVED** [review]

File: lib/job-runner/tests/basic.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	assert_eq!(0, setup.child.wait().unwrap().code().unwrap());
+	assert_eq!("starting", setup.msg_rx.recv().unwrap().message);
+	assert_eq!("exiting", setup.msg_rx.recv().unwrap().message);
+	assert!(setup.msg_rx.recv().is_err());
+}
+
+#[test]
+fn exit_code() {
+	let mut setup = common::setup("echo starting; exit 69");
+	assert_eq!(69, setup.child.wait().unwrap().code().unwrap());
```
Comment (AngelOnFira) [inline lib/job-runner/tests/basic.rs:15]: [nicerocket](https://app.graphite.dev/macros/rivet-gg)

**Review by AngelOnFira — APPROVED** [review]

**Pull request merged**